firedrake/supermeshimpl.pyx

```diff
@@ -98,3 +103,50 @@ def assemble_mixed_mass_matrix(V_A, V_B, candidates,
 
     CHKERR(MatAssemblyBegin(mat.mat, MAT_FINAL_ASSEMBLY))
     CHKERR(MatAssemblyEnd(mat.mat, MAT_FINAL_ASSEMBLY))
+
+
+def intersection_finder(mesh_A, mesh_B):
+    # Plan:
+    # Call libsupermesh_sort_intersection_finder_set_input
+    # Call libsupermesh_sort_intersection_finder_query_output
+    # Call libsupermesh_sort_intersection_finder_get_output
+    # Return the output
+
+    cdef:
+        numpy.ndarray[long, ndim=2, mode="c"] vertex_map_A, vertex_map_B
+        numpy.ndarray[double, ndim=2, mode="c"] vertices_A, vertices_B
+        long nindices
+        numpy.ndarray[long, ndim=1, mode="c"] indices, indptr
+        long nnodes_A, nnodes_B, ncells_A, ncells_B
+        int dim_A, dim_B, loc_A, loc_B
+
+    vertices_A = mesh_A.coordinates.dat.data_ro_with_halos
+    vertices_B = mesh_B.coordinates.dat.data_ro_with_halos
+    vertex_map_A = mesh_A.coordinates.cell_node_map().values_with_halo.astype(long)
+    vertex_map_B = mesh_B.coordinates.cell_node_map().values_with_halo.astype(long)
+    nnodes_A = mesh_A.num_vertices()
+    nnodes_B = mesh_B.num_vertices()
+    dim_A = mesh_A.geometric_dimension()
+    dim_B = mesh_B.geometric_dimension()
+    ncells_A = mesh_A.num_cells()
+    ncells_B = mesh_B.num_cells()
+    loc_A = vertex_map_A.shape[1]
+    loc_B = vertex_map_B.shape[1]
+
+    libsupermesh_tree_intersection_finder_set_input(&nnodes_A, &dim_A, &ncells_A, &loc_A,
+                                                    &nnodes_B, &dim_B, &ncells_B, &loc_B,
+                                                    <const double*>vertices_A.data, <const long*>vertex_map_A.data, <const double*>vertices_B.data, <const long*>vertex_map_B.data)
+
+    libsupermesh_tree_intersection_finder_query_output(&nindices)
+
+    indices = numpy.empty((nindices,), dtype=long)
+    indptr  = numpy.empty((mesh_A.num_cells() + 1,), dtype=long)
+
+    libsupermesh_tree_intersection_finder_get_output(&ncells_A, &nindices, <long*>indices.data, <long*>indptr.data)
```

Comment:
What assumptions are there on the parallel decomposition of mesh_A and mesh_B. What happens if they don't coincide? Is the oracle still correct?

If it is not correct, what is the outcome? Silently wrong answer? Error condition?

Reply:
Possible option: provide metadata in the form of a weak reference to mesh_A and mesh_B in a magic slot that says "I guarantee that mesh_B has a decomposition that matches mesh_A". By default, this slot is empty => error. Pragmatic (or whatever) interfacing the remeshing can fill this slot in mesh_B with a weakref to mesh_A and we check this and satisfy ourselves that this is OK.

Reply:
This also fixes the communicator issue above.

Reply:
(But leave those checks in).

Reply:
There is a (strong) constraint on the parallel decomposition: each process needs to see all relevant source data for the computation of its piece of the target.

I don't know how to implement the magic slot idea you describe. Could you expand a little?

Reply:
I have pushed a first implementation of the magic slot design you suggest.